Adoption design (#2067)
Azure/azure-service-operator: docs/hugo/content/architecture-decision-records/ADR-2022-02-Adoption-Policy.md

# Adoption Policy 

## Context

Resources which exist in Azure prior to being created in Kubernetes are "adopted" by the operator when the resource
is created in Kubernetes. Today this adoption process is automatic and implicit. As a user there isn't an easy way
to tell if a resource you're about to `kubectl apply` already exists in Azure (will be adopted) or does not exist in
Azure (will be created).

We have recently proposed support for [reconcile-policy](https://github.com/Azure/azure-service-operator/pull/2060) which
is related to adoption. Some of `reconcile-policy`'s primary use cases are for when the user does not want ASO to
fully manage an adopted resource.

The following are principles we should follow related to resource adoption:

1. **The operator should adopt resources by default:** This is standard goal-seeking/desired-state behavior. 
   If you create a resource in Kubernetes that says: "I expect there to be a Virtual Machine that looks like this
   at location X" and that Virtual Machine already exists, then we are in the desired state.

>    The alternative would be to error if the resource already exists, which goes against the Kubernetes model.
> 2. **Users must be able to instruct the operator to stop making changes to the backing Azure resource:**

Reviewer: Suggest a blank line between each principle, since they're paragraphs.

```suggestion
   The alternative would be to error if the resource already exists, which goes against the Kubernetes model.
   
2. **Users must be able to instruct the operator to stop making changes to the backing Azure resource:**
```

   This is enabled by `reconcile-policy`.

   This is useful for the following scenarios:

>    - The operator is doing something wrong and the user needs an out.

Reviewer: Not sure of the phrasing here, but feels too informal at the moment.

```suggestion
   - The operator is doing something wrong and the user needs a way to stop it from proceeding
```

   - Pausing the operator temporarily while they make some manual fix via the Azure portal.
   - Creating a resource in ASO purely to get ownership references to work. The resource in question is not managed
     by ASO at all. It was created and managed some other way (portal, terraform, az-cli, etc).
3. **Users must be able to instruct the operator to skip deletion of the backing resource:**

>    This is useful for scenarios such as moving ownership of resources from ASO in cluster A to ASO in cluster B.

Reviewer: ```suggestion
   This is useful for scenarios such as moving ownership of resources from ASO in cluster A to ASO in cluster B,
   or as a safety net to protect stateful resources such as databases from accidental deletion.
```

Author: Added

   This is enabled by `reconcile-policy`.

### Open question: what should the default `reconcile-policy` be for resources which are adopted?

There are two main options.

#### Option 1: The default `reconcile-policy` for adopted resources is the same as for resources ASO created.

ASO automatically adopts, updates, and will delete backing Azure resources as normal unless the user
explicitly sets `reconcile-policy`.

**Advantages:**
- It's easy to explain what the behavior is. Simplistic users don't have to understand `reconcile-policy` (they won't
  ever see it unless they set it themselves).

**Disadvantages:**
- Users have no good way to know their resource existed previously. They may `kubectl delete` it without realizing that
  other things depend on this resource. Deletion (and some mutation) is destructive in Azure so lost data, etc cannot 

>   easily be recovered.

Reviewer 2: I think this is a minor issue. Once I decide as a user to use ASO as the controller over my resources, I expect ASO to own the resources and don't care so much which already existed before and which didn't.
If I do care than I'll be explicit about it and set the `reconcile-policy` label, but then at least it's obvious in the source code of the resource (whether helm chart or bare yaml) that I for instance wanted to abandon a resource or skip the reconcile on it.

Reviewer: I agree this is the case for users who understand the service operator model.

I have a concern about users who aren't experienced/knowledgable.

Consider what happens if a potential user starts auditioning ASO to see if it meets their needs and then inadvertently deletes something in Azure that they really shouldn't have? The most likely case would be a database or similar resource in their Dev/Test subscription. Worst case, they delete something from prod.

Is it really useful at that point for us to say "You should have understood what you were doing before you did it?"

While this is a minor issue overall, it wouldn't feel minor to that particular user.

I'm interested to see if there's a safe default that we can reasonably adopt, so that data-loss is less likely.

Reviewer 3: Another important use case to consider is "update". For example, what happens if I create a LB with the name of a LB that already exists (without realizing it) but specify properties (eg. LB rules) that are different from the existing LB? Will ASO silently update my LB, potentially breaking other applications that rely on that resource? I can see how a "kubectl delete" is an explicit action but creating a resource with a name collision might be more surprising. 

A way to look at it another way is that performing a PUT in Azure on an existing resource does result in the resource being updated in the same pattern.

Author: > For example, what happens if I create a LB with the name of a LB that already exists (without realizing it) but specify properties (eg. LB rules) that are different from the existing LB?

Agree that this is definitely an important case to consider. There's nothing currently in this proposal that would automatically block mutation of an existing resource. In all cases that I could find, the mutations are reversible. I was thinking that something like shrinking PostgreSQL DB size would be non-reversible, but they disallow that operation (possibly because it's destructive). In your example, assuming you know what your old LB rules are, you can always put them back later, so at least there's an "out". The focus was on the delete case because there is a possibility that there's no recovery if you delete your database.

Option 2 outlined here would help with the delete case but does nothing for the update case. I don't see a way to satisfy this goal:
> **The operator should adopt resources by default:** This is standard goal-seeking/desired-state behavior. 
   If you create a resource in Kubernetes that says: "I expect there to be a Virtual Machine that looks like this
   at location X" and that Virtual Machine already exists, then we are in the desired state.
   The alternative would be to error if the resource already exists, which goes against the Kubernetes model.

and _also_ make adoption truly "safe". As you say, it's like submitting an ARM template for an existing resource. If you do it, it will modify the resource.

Author: I have amended this design to talk about two different questions here:
1. What is the default behavior?
2. Can users override that behavior?

My proposed choice is that the default behavior is to do what we're doing today (adopt resources if they already existed and manage them), _but_ unlike the original design (from nearly a year ago!), I also suggest we allow per-resource and operator-level overrides to this behavior so that if there are particular resources where users want a different behavior they can select that behavior.

Resolving this conversation as most of the points brought up in it have been incorporated into the design.

**Thoughts/comments:**

This sort-of matches the behavior of ARM templates and SDKs. If you call the `DELETE` API the resource is gone.

Some discussion of this topic has happened over in [ACK](https://github.com/aws-controllers-k8s/community/issues/82).

[Crossplane](https://github.com/crossplane/crossplane/issues/2590) doesn't support any notion of `reconcile-policy` at all.
`kubectl delete` always deletes the backing resource no matter what.

[GCP](https://cloud.google.com/config-connector/docs/how-to/managing-deleting-resources#creating_a_resource) does it this way:
> When you create a resource, Config Connector creates the resource if it doesn't exist. If a Google Cloud resource 
> already exists with the same name, then Config Connector acquires the resource and manages it.

GCP also [supports](https://cloud.google.com/config-connector/docs/how-to/managing-deleting-resources) a 
`reconcile-policy`-like annotation, so doing it like this would basically be doing it exactly like GCP.

> Kuberentes has a similar concept for `StorageClasses`/`PersistentVolumes` called 

Reviewer 3: ```suggestion
Kubernetes has a similar concept for `StorageClasses`/`PersistentVolumes` called 
```

> [reclaimPolicy](https://kubernetes.io/docs/concepts/storage/storage-classes/#reclaim-policy), whose default

Author: Clarified this in the docs

seems to be `Delete`.

#### Option 2: The default `reconcile-policy` for adopted resource is `abandon-on-delete`.

ASO detects that a resource previously existed and sets `reconcile-policy` automatically to `abandon-on-delete`.
A new option in the ASO config `adoption-policy` would control this default so users could opt into the Option 1 style
discussed above if they wanted (but it's not the default behavior).

**Advantages:**

> - More flexibility. Users can choose which behavior they want when deploying ASO. 

Reviewer 2: Hmm correct me if I'm wrong, but I don't see this as a distinctive advantage to this approach. Users should be just as flexible with the other approach as they can actively choose and set the reconcile-policy. This would just be to attempt a better default, an attempt in guessing what the user actually wants without them actively setting it.

Author: What I was getting at with flexibility here was that, if users wanted the operator to autodetect adopted resources, it could. If they didn't want that, they can just set `adoption-policy` and get the option1 behavior. We'd basically be letting them choose which behavior they want.

Whereas if we just go with option1 and no `adoption-policy` at the operator level then the users can't have the operator "know" for them, they'll have to know and annotate things accordingly.

- Defaults to a more conservative `reconcile-policy` to protect them from accidentally deleting things they may not
  have wanted to delete.

> **Disadvantages:**

Reviewer 2: I think we can add one more to this list which is potential inconsistency across resources.
Imagine you have a helm chart of 10 ASO resources, 5 of which were already present on Azure, and the 5 others are just introduced through the helm chart. Now half is 'owned' by the ASO controller and the other half is not. You forget about that and a long time later you run `helm uninstall` because you intend to get rid of all the resources in the chart. And in fact even all the resources get deleted from the cluster and you won't see them anymore. But half of those resources would still continue to remain in Azure.

Author: I covered this I think with this line:
> Bias towards safety may cause resource user thinks was deleted to continue existing (and billing) in Azure. Not a data
  loss though.

- Bias towards safety may cause resource user thinks was deleted to continue existing (and billing) in Azure. Not a data
  loss though.
- Implementation complexity: We would have to issue some extra requests to Azure to detect if the resource exists before `PUT`-ing it. 
  
  Since `reconcile-policy` as an annotation isn't managed solely by the user or the operator, deciding if
  we should trigger reconciliation events on its change is tricky (we don't want our defaulting of it to trigger an event). 
  It's doable though.

**Thoughts/comments:**

There's some discussion of a similar feature over in [ACK](https://github.com/aws-controllers-k8s/community/pull/1148#discussion_r794605821).
Jay Pipes had an interesting comment:
> In the case of ACK and the user experience of resource deletion, I posit that the behaviour of our interface has 
> no surprises and is actually hard to misuse. It doesn't have surprises because our interface to delete a resource 
> is to call kubectl delete on that resource. This is as non-surprising as it gets. It is hard to misuse because 
> calling kubectl delete is an explicit action the user has to take and it's not a confusing verb or semantic. 

> > "delete" means to remove the resource.

Reviewer 2: fully agree with this.

Reviewer: I fully agree that there's no ambiguity that `kubectl delete` means remove the resource _from the cluster_.

That said, I do think that different users will have different expectations about whether deletion from the cluster necessarily requires deletion from Azure as a natural consequence.
